### PR TITLE
EVG-20744 Add unique senders for github apps

### PR DESCRIPTION
**Pull request by bynn**
EVG-20744

### Description
GitHub apps don't support cross-org authentication so you need to generate a new access token for every owner. 
This change makes senders for each owner and caches it to the env.
GitHub app tokens expire after 1 hour so we re-create the token and sender after it expires 

### Testing
Can't test on staging as the staging github app is only installed on 1 org so I can't test the cross-org authing 
But confirmed that the fallback works and we use the legacy global sender when the app fails to create a sender

### CODE REVIEW

**Review by ablack12 — COMMENTED** [review]
To verify, were you able to test that this _does_ work for different repos in the org, at least? (like both evergreen and commit-queue-sandbox?)

File: environment.go [hunk 1]
```diff
@@ -1057,6 +1063,46 @@ func (e *envState) SaveConfig(ctx context.Context) error {
 	return errors.WithStack(UpdateConfig(ctx, &copy))
 }
 
+func (e *envState) GetGitHubSender(owner, repo string) (send.Sender, error) {
```
[the remark this answers is not in the document]
Comment (ablack12) [reply environment.go:1066]: And I would include a comment about how the repo is only used for generating, not needed for token uniqueness

[the remark this answers is not in the document]
Comment (bynn) [reply environment.go:1074]: done

**Review by bynn — COMMENTED** [review]
Yes, I tested the changes with 2 different repos but will test it again with the changes I made here

File: environment.go [hunk 2]
```diff
@@ -267,6 +273,11 @@ type closerOp struct {
 	closerFn   func(context.Context) error
 }
 
+type cachedSender struct {
```
[the remark this answers is not in the document]
Comment (ablack12) [reply environment.go:279]: Perhaps we should call this cachedGithubSender (or just githubSender)

File: environment.go [hunk 3]
```diff
@@ -1057,6 +1063,46 @@ func (e *envState) SaveConfig(ctx context.Context) error {
...
+		token, err := e.settings.CreateInstallationToken(e.ctx, owner, repo, nil)
+		if err != nil {
+			// TODO EVG-19966: Delete fallback to legacy GitHub sender
+			grip.Debug(message.WrapError(err, message.Fields{
+				"message": "error creating installation token for GitHub sender",
+				"owner":   owner,
+				"repo":    repo,
+				"ticket":  "EVG-19966",
+			}))
+			return e.GetSender(SenderGithubStatus)
```
[the remark this answers is not in the document]
Comment (ablack12) [reply environment.go]: oh oof i hadn't even thought of that.

File: units/github_status_api.go [hunk 4]
```diff
@@ -255,6 +246,13 @@ func (j *githubStatusUpdateJob) fetch() (*message.GithubStatus, error) {
 	return &status, nil
 }
 
+func (j *githubStatusUpdateJob) setSender(owner, repo string) error {
+	// TODO EVG-19966: Delete fallback to legacy GitHub sender
```
Comment (ablack12) [inline units/github_status_api.go:250]: Do we need this todo here? this doesn't itself call any fallback logic

File: units/github_status_refresh.go [hunk 5]
```diff
@@ -96,12 +96,10 @@ func (j *githubStatusRefreshJob) fetch(ctx context.Context) error {
 	if j.urlBase == "" {
 		return errors.New("url base doesn't exist")
 	}
-	if j.sender == nil {
-		var err error
-		j.sender, err = j.env.GetSender(evergreen.SenderGithubStatus)
-		if err != nil {
-			return err
-		}
+	// TODO EVG-19966: Delete fallback to legacy GitHub sender
```
Comment (ablack12) [inline units/github_status_refresh.go:99]: same here

File: model/patch_lifecycle.go [hunk 6]
```diff
@@ -1267,7 +1267,8 @@ func SendCommitQueueResult(ctx context.Context, p *patch.Patch, status message.G
 		Description: description,
 		URL:         url,
 	}
-	sender, err := evergreen.GetEnvironment().GetSender(evergreen.SenderGithubStatus)
+	// TODO EVG-19966: Delete fallback to legacy GitHub sender
```
Comment (ablack12) [inline model/patch_lifecycle.go:1270]: same here

File: environment.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ablack12) [reply environment.go]: I more meant, should we be initializing the legacy sender _also_ in the new map as opposed to the old map, but that does seem chaotic and I'm fine with breaking it up.

**Review by ablack12 — APPROVED** [review]
LGTM with comment nits

**Pull request merged**